No clue  at Problem 3

SGJJSR opened the pull request:
see in text comments

SGJJSR committed:
No clue @Problem 3

see in text comments

nicolaslara commented:
dude! Add `.py` to your python files!!! haha (`gis_exercise_2.py`)

nicolaslara commented:
Also curious... which text editor are you using?

nicolaslara committed:
cimmenting out code that doesn't work

nicolaslara commented on gis_exercise_2:
```diff
@@ -18,7 +18,75 @@
 
-# Problem 2: Points to map 
+import os
+os.chdir("C:\\.\\Exercise_2")
```
Do not do `chdir` inside your code. Unless you're writing system code, you probably don't need to do this (and you definitely don't need to do it here). Your code should work from the directory you call it in and all your paths should be relative to your project's "root" directory.

What is the purpose of changing the directory here?

SGJJSR replied:
1. Aw shit I thought I remembered this time! :D so sry! I really need to remember these things.

2. I am using Python notebook for coding and just a normal text editor (notepad) or csv for the data I import..not sure what you mean with the text editor

3. I think I just wanted to check it out... I thought of it more like the directory in STATA which just specifies where my input/output goes to. So what you say, I should definitely have the .py file at the same location where also all my data / graphs input output goes to? Why is this so mandatory? Either way, then I start doing it!

nicolaslara replied:
> Aw shit I thought I remembered this time! :D so sry! I really need to remember these things.

haha, no worries, I asked cause there may be a reason I don't know about

> I am using Python notebook for coding and just a normal text editor (notepad) or csv for the data I import..not sure what you mean with the text editor

That makes sense. A notebook is common in scientific programming, so it's good you get familiar with them. However, in "regular" programming you write text files (hence a text editor) and "run" them on the shell. Discussions about which text editor is better are HUGE amongst programmers, haha.

Anyway, if you want to try a text editor to edit your files (so that it highlights the syntax, errors, etc) you can try something simple like Sublime or Atom... Or an IDE, like PyCharm or VSCode. I use PyCharm a lot, and I'd suggest you try it. Just the error highlighting helps a lot to see when something is wrong. 

> So what you say, I should definitely have the .py file at the same location where also all my data / graphs input output goes to? 

Definitely is a strong word. It all depends on what you're trying to do. But I'd say it's a "best practice" to do so.

I'd say a simple solution is to add a directory for the data, and another one for outputs. I did something like that in my pull request: https://github.com/SGJJSR/Python_challenges/pull/8 

> Why is this so mandatory? 

Again, I wouldn't say mandatory, but It's just more organized, not a lot of files everywhere. For example, you can delete everything inside the outputs directory, cause you know that they will be generated by the script. It also helps when other people (me, in this case) is going to look at your code. Think of it this way: if your code was going to be on the internet and you or someone else was going to have to pick up where you left off in a couple of years... what would you like for them to find so that they can start working as soon as possible?

There's a famous saying that says:

> "Always code as if the guy who ends up reading your code will be a violent psychopath who knows where you live."

nicolaslara commented:
Let me review what's going on here and point you in the right direction, cause I know that's what you care about the most, but once you're done with the current problem... Let's take a moment to clean up your code and make it follow best practices. That way you'll also learn a bit about how to structure your code and work with others (which is really important!)

Some of the stuff you should review (notes to self so I can remember later):

 * Your code should run unless "not running" is the issue you're trying to solve.
 * You should have a `requirements.txt` file with all your dependencies so that others can easily install them and run your code
 * You should provide all the necessary data or instructions about how to obtain it.
 * Your paths should be relative and not OS dependent (example `./output/result.txt` instead of `C:\\Users\...`
 * Are you familiar with `ipython` and `ipdb`?

SGJJSR committed:
Rename gis_exercise_2 to gis_exercise_2.py

nicolaslara committed:
removed absolute paths so the code can run

nicolaslara committed:
added the necessary data

nicolaslara committed:
added a requirements file so that the requirements can be installed by others

nicolaslara committed:
using the data from ./data/

nicolaslara committed:
Merge branch 'SGJJSR-exercise_2.py_problems' of github.com:SGJJSR/Python_challenges into SGJJSR-exercise_2.py_problems

nicolaslara committed:
fixed slow iteration

nicolaslara committed:
removed undefined variable. Replaced by one that exists. Is this the correct one?

nicolaslara commented:
I think your problem here is that you're trying to continue solving the problem without taking the time to step back and understand what each part of the code is doing (i.e.: throwing some code at the wall to see what sticks). You're using `groupby`, `sort_values` and `apply`. I suggest you spend some time trying to understand what each does individually in an environment isolated from this project (a test pandas dataframe). This is a good resource for groupby: https://realpython.com/pandas-groupby/#how-pandas-groupby-works. This one too: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html.

In short: GroupBy by itself is not enough, to apply a function to the groupped elements. So if you have two groups, A and B, with elements 1, 2 and 3 for each group. Doing `df.groupby(['group_name']).sum()` would sum the numbers for each element in the group (A->6, b-> 6). You can try this by manually creating simple dataframes and manipulating them.

--- 

It isn't clear from the problem description which of the following is the problem you're trying to solve:

 * The distance between the firs and last point for a user (Final distance travelled)
 * The distance between each pair of points (Total distance travelled)

Independently of which one you choose, there are many ways to write code that solves your problem. Some may be one line, others may be a for loop, or storing some temporary elements and manipulating them; but the general process for obtaining the result is probably similar.

Assuming the problem is the second one, the process will be like the follwong "pseudocode":

 * Consider the list of entries in the dataframe that have the same user id
 * Order them by timestamp
 * For each pair of elements in that list, calculate the distance between them
 * Sum all the distances for that user.

If it's the first one:

 * Consider the list of entries in the dataframe that have the same user id
 * Find the one with the lowest timestamp and the highest timestamp
 * Calculate the distance between those two points

I'd recommend: Take a step back, play with the functions that you're trying to use so you understand them, and then try to solve the problem step by step. 

If that doesn't work, you can also create an analogous problem with your own sample data (by writing a dataframe with 4-5 rows manually) and try to solve it there. That way you can see what each step is doing and if it's giving you the result you expect.

If you're still lost tomorrow, I can show you some solutions and explain why they work.

SGJJSR committed:
Merge pull request #8 from nicolaslara/SGJJSR-exercise_2.py_problems

Initial fixes and reorganization

nicolaslara committed:
Merge branch 'SGJJSR-exercise_2.py_problems' of github.com:SGJJSR/Python_challenges into SGJJSR-exercise_2.py_problems

SGJJSR committed:
Merge pull request #9 from nicolaslara/SGJJSR-exercise_2.py_problems

The changes on the loop

nicolaslara commented:
BTW, I also remembered: I have an implementation of the "Total distance travelled" calculation which I haven't pushed so I wouldn't spoil it for you. If you wanna take a look, let me know and I'll push it.

SGJJSR commented:
ok thanks!

I figured out how to sort the values according to the time, but I
struggle with the "combine" part and merging it into a geopandaframe on
which I could perform LineString actions..
But I didn't have much time these last days, so I will give it another shot
tonight! I guess by tomorrow, I would need some help so you could push it
by then :D

Am Mo., 11. Mai 2020 um 14:07 Uhr schrieb Nicolas Lara <
notifications@github.com>:

> BTW, I also remembered: I have an implementation of the "Total distance
> travelled" calculation which I haven't pushed so I wouldn't spoil it for
> you. If you wanna take a look, let me know and I'll push it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SGJJSR/Python_challenges/pull/7#issuecomment-626661311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AND7PNVL55GHHIGXESI3CH3RQ7S6ZANCNFSM4M4CIICQ>
> .
>

nicolaslara committed:
calculating distance

SGJJSR committed:
Merge pull request #10 from nicolaslara/SGJJSR-exercise_2.py_problems

calculating distance

SGJJSR committed:
Finish gis_exercise_2.py

I just toyed around a little bit, although not everything seems to be fine.
I also downloaded pycharm, but need to familiarize me with it still (it does need a lot CPU right? Maybe it's not the best for my old laptop)

SGJJSR committed:
distinguished between comment and task

## = Task & given by the exercise
# = comment

SGJJSR commented on gis_exercise_2.py:
```diff
@@ -121,8 +121,12 @@
+#    kruger_park.loc[index, 'geometry'] = Point((kruger_park.loc[index, 'lon']), (kruger_park.loc[index, 'lat']))
+
+# this is a faster way of doing it. Not sure if it's needed. I'll review that later.
+kruger_park['geometry'] = [Point(x, y) for x,y in kruger_park[['lat', 'lon']].values]
```
Man, I feel really stupid...

I have been following the course of 2018 and I just noticed today that there is also [one for 2019](https://automating-gis-processes.github.io/site/lessons/L2/exercise-2.html) available.
The one from 2019 explains so much of what I was pondering about and what I did wrong :D

oh well. better late than never I guess haha
